app/controllers/application_controller.rb

```diff
@@ -1,10 +1,146 @@
+require 'pry'
 require './config/environment'
 
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
   end
 
-end
\ No newline at end of file
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect to '/tweets'
+    end
+    erb :'users/create_user'
+  end
+
+  post '/signup' do
+    @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+    if @user.save
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    elsif @user.username.nil?
+      redirect '/signup'
+    elsif @user.email.nil?
+      redirect '/signup'
+    elsif @user.password.nil?
+      redirect '/signup'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/tweets' do
+    if !logged_in?
+      redirect '/login'
+    else
+    @user = current_user(session)
+    @tweets = Tweet.all
+    erb :'tweets/tweets'
+  end
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect to '/tweets'
+    else
+    erb :'users/login'
+  end
+end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    session[:user_id] = @user.id
+    redirect to '/tweets'
+  end
+
+  get '/logout' do
+    if !logged_in?
+      redirect '/'
+    else
+    session.clear
+    redirect '/login'
+  end
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+
+  get '/tweets/new' do
+    if !logged_in?
+      redirect to '/login'
+    else
+      erb :'tweets/create_tweet'
+    end
+  end
+
+  post '/tweets' do
+    @user = current_user(session)
+    if params[:content] != ""
+    @tweet = Tweet.create(content: params[:content], user_id: @user.id)
+    redirect to '/tweets'
+    else
+      redirect to '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if !logged_in?
+      redirect to '/login'
+    else
+    @tweet = Tweet.find_by_id(params[:id])
+    erb :'tweets/show_tweet'
+  end
+  end
+
+  get '/tweets/:id/edit' do
+    if !logged_in?
+      redirect to '/login'
+    else
+    @tweet = Tweet.find_by_id(params[:id])
+    erb :'tweets/edit_tweet'
+  end
+end
+
+
+  patch '/tweets/:id/edit' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if params[:content] != ""
+      @tweet.update(content: params[:content])
+      redirect to '/tweets'
+    else
+    redirect to "/tweets/#{@tweet.id}/edit"
+  end
+end
+
+
+delete '/tweets/:id/delete' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if logged_in? && current_user(session).id == @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.